### ISSUE TITLE
Multiple values in a parameter

### ISSUE THREAD

**Issue by marvinm2** [CLARIAH/grlc#211]
Is there a way to define lists of parameters, or somehow combine the different parameters with VALUES? When I start putting parameters in those sections of the query, the parameters are not present in the API that is created. 

I also tried using a single parameter(with multiple variables in it) and putting it directly in the VALUES part of the query, but that also doesn't work.

The query that I am trying to use: 
https://github.com/marvinm2/AOPWikiQueries/blob/master/get-methods-for-multiple-aops.rq

Basically this should provide the methods for multiple AOPs, as done with this query for a single one:
https://github.com/marvinm2/AOPWikiQueries/blob/master/get-methods-for-aop.rq

I hope that my question is a bit clear. I am not too familiar with the terminology.

The variables that should be used in this query can be, for example: 
http://identifiers.org/aop/3    (or aop:3 if directly in the SPARQL endpoint)
http://identifiers.org/aop/12   (or aop:12 if directly in the SPARQL endpoint)

**Comment by c-martinez** [comment]
Hi @marvinm2,

I am not sure if I understand your question correctly, but as I understand, you would like to select an AOP value from a list of available AOPs. In that case, the `#+ enumerate` decorator is what you are looking for. See if this query is more or less what you are after:

```
#+ summary: Get methods for all Key Events of an Adverse Outcome Pathway. For the variable of aopfilter, fill the complete URI for the AOP. This consists of "http://identifiers.org/aop/" + the AOP ID of interest. In this example, this was AOP 3. Therefore, the variable aopfilter is filled with "http://identifiers.org/aop/3".
#+ endpoint_in_url: False
#+ tags:
#+   - Methods
#+ defaults:
#+   - aopfilter: http://identifiers.org/aop/3
#+ enumerate:
#+   - aopfilter:
#+     - http://identifiers.org/aop/3
#+     - http://identifiers.org/aop/12

prefix dcterms: <http://purl.org/dc/terms/>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix aop: <http://identifiers.org/aop/>
prefix aopo: <http://aopkb.org/aop_ontology#>
prefix mmo: <http://purl.obolibrary.org/obo/MMO_>

select  ?AopLabel?KeLookUp ?AssayText
where {
 ?KeLook a aopo:KeyEvent ;
 rdfs:label ?KeLookUp ;
 dcterms:isPartOf  ?_aopfilter_iri;
 mmo:0000000 ?AssayText .
 ?_aopfilter_iri rdfs:label ?AopLabel .
}
````
Hope this helps!

**Comment by marvinm2** [comment]
Thank you for the response! However, I am actually looking for something different: I would like to perform this query in 1 time for both of these AOPs at the same time, like a batch call. So that 1 query can give me the full table of the methods (outcome of the query) for all of the AOPs of interest. Sorry for not being clear before. 

I would like to do this: like the outcome of this query if you fill it in http://aopwiki-rdf.prod.openrisknet.org/sparql/: 

```sparql
SELECT  ?AopLabel?KeLookUp ?AssayText
WHERE {
 VALUES ?aops {aop:3 aop:12}
 ?KeLook a aopo:KeyEvent ;
 rdfs:label ?KeLookUp ;
 dcterms:isPartOf  ?aops; 
 mmo:0000000 ?AssayText .
 ?aops rdfs:label ?AopLabel .     
} ORDER BY ?AopLabel
```

If this kind of list-queries/batch-queries is possible, I could use such constructs in many of my queries where single calls for multiple parameters are very valuable, for example with chemical mixtures (querying for multiple chemical compounds at once).

**Comment by c-martinez** [comment]
Hi @marvinm2 -- thanks for clarifying -- I am not sure if this is possible. I was trying to make it work by using a `BIND` clause, but I don't think you can inject a `?_var` that way. 

@albertmeronyo -- do you have any ideas how this could work?

**Comment by c-martinez** [comment]
Hi @marvinm2 -- I think it might be possible to do what you are after, by using SPARQL functionality -- but this would depend on your endpoint. grlc would give you an API on top of this.

For instance, I tried the following query (using Apache Jena as triplestore):
```SPARQL
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
PREFIX apf: <http://jena.apache.org/ARQ/property#>

SELECT DISTINCT ?id ?name
WHERE {
  BIND( iri(?id_str) as ?id) .
  ?id_str apf:strSplit ("http://example.org/#spiderman,http://example.org/#green-goblin>" ",") .
  ?id foaf:name ?name
}
```
This query gives me an ID and a name for each of the iri's in my string `"http://example.org/#spiderman,http://example.org/#green-goblin>"`. You can replace that for a variable: 
```SPARQL
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
PREFIX apf: <http://jena.apache.org/ARQ/property#>

SELECT DISTINCT ?id ?name
WHERE {
  BIND( iri(?id_str) as ?id) .
  ?id_str apf:strSplit (?_names ",") .
  ?id foaf:name ?name
}
```
Then you could pass `?_names` with however many iri's as you need. How to do this specifically, would depend on your triplestore, I'm afraid.

I hope this helps!

Since this is not grlc specific, I'll close this issue.
